File: src/main/org/nlogo/prim/_linkwith.java
package org.nlogo.prim;

import org.nlogo.agent.AgentSet;
import org.nlogo.agent.Link;
import org.nlogo.agent.Turtle;
import org.nlogo.api.LogoException;
import org.nlogo.nvm.Reporter;
import org.nlogo.nvm.Syntax;

public final strictfp class _linkwith
    extends Reporter {
  private final String breedName;

  public _linkwith() {
    breedName = null;
  }

  public _linkwith(String breedName) {
    this.breedName = breedName;
  }

  @Override
  public Syntax syntax() {
    int[] right = {Syntax.TYPE_AGENT};
    int ret = Syntax.TYPE_LINK;
    return Syntax.reporterSyntax(right, ret, "-T--");
  }

  @Override
  public String toString() {
    return super.toString() + ":" + breedName;
  }

  @Override
  public Object report(final org.nlogo.nvm.Context context) throws LogoException {
    org.nlogo.agent.LinkManager linkManager = world.linkManager;
    Turtle parent = (Turtle) context.agent;
    Turtle target = argEvalTurtle(context, 0);
    AgentSet breed = breedName == null ? world.links() : world.getLinkBreed(breedName);
    mustNotBeDirected(breed, context);
    Link link = linkManager.findLinkEitherWay(parent, target, breed, true);
    if (link == null) {
      return org.nlogo.api.Nobody.NOBODY;
    }
    return link;
  }
}

File: src/main/prim/etc/_linkwith.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.core.{ Syntax, SyntaxJ }
import org.nlogo.api.Nobody
import org.nlogo.agent.{ Turtle, LinkManager }
import org.nlogo.nvm.{ Reporter, Context, EngineException }

class _linkwith(breedName: String) extends Reporter {

  def this() = this(null)

  override def syntax =
    SyntaxJ.reporterSyntax(
      Array(Syntax.AgentType),
      Syntax.LinkType, "-T--")

  override def toString =
    super.toString + ":" + breedName

  override def report(context: Context): AnyRef = {
    val parent = context.agent.asInstanceOf[Turtle]
    val target = argEvalTurtle(context, 0)
    val breed =
      if (breedName == null)
        world.links
      else
        world.getLinkBreed(breedName)
    for(err <- LinkManager.mustNotBeDirected(breed))
      throw new EngineException(context, this, err)
    val link = world.linkManager.findLinkEitherWay(parent, target, breed, true)
    if (link == null)
      Nobody
    else
      link
  }

}
